File: php/invoices.php
<?php

require_once 'entity.php';
require_once 'jobs.php';
require_once 'payments.php';
require_once 'items.php';
require_once 'addresses.php';
require_once 'domains.php';
require_once 'users.php';
require_once 'files.php';
require_once 'companies.php';
require_once 'phone_numbers.php';
require_once 'email_addresses.php';
require_once './../php_aux/invoice_status.php';
require_once './../php_aux/money_protocol.php';

class Invoice extends Entity
{
    public static $resource = '/invoices/';
    public static $json_name = 'invoice';

    public function __construct()
    {
        $this->json_property('id', 'integer');
        $this->json_property('creation_date', 'DateTime');
        $this->json_property('payment_deadline', 'DateTime');
        $this->json_property('reminded', 'DateTime');
        $this->json_property('note', 'string');
        $this->json_property('send_sms','boolean');
        $this->json_property('send_email','boolean');
        $this->json_property('accept_stripe', 'boolean');
        $this->json_property('accept_paypal', 'boolean');
        $this->json_property('unpaid', 'boolean');
        $this->json_property('accept_bank_transfer', 'boolean');
        $this->json_property('accept_phone_payment', 'boolean');
        $this->json_property('is_remindable', 'boolean');
        $this->json_property('force_reminders', 'boolean');
        $this->json_property('total_cost', 'float');
        $this->json_property('subtotal_cost', 'float');
        $this->json_property('tax_amount', 'float');
        $this->json_property('currency', 'string');
        $this->json_property('invoice_token', 'string');
        $this->json_property('reminder_message', 'string');
        $this->json_property('responsible_manager','User', $many = False,
                             $recursive = True);
        $this->json_property('creator','User', $many = False,
                             $recursive = True);
        $this->json_property('client','User', $many = False,
                             $recursive = True);
        $this->json_property('client_phone','PhoneNumber', $many = False,
                             $recursive = True);
        $this->json_property('client_email','EmailAddress', $many = False,
                             $recursive = True);
        $this->json_property('jobs','Job', $many = True,
                             $recursive = True);
        $this->json_property('client_company','Company', $many = False,
                             $recursive = True);
        $this->json_property('client_company_phone','PhoneNumber', $many = False,
                             $recursive = True);
        $this->json_property('client_company_email','EmailAddress', $many = False,
                             $recursive = True);
        $this->json_property('items','Item', $many = True,
                             $recursive = True);
        $this->json_property('shipping','Address', $many = False,
                             $recursive = True);
        $this->json_property('pdf','File', $many = False,
                             $recursive = True);
        $this->json_property('receipt','File', $many = False,
                             $recursive = True);
        $this->json_property('domain', 'Domain', $many = False,
                             $default = '1', $recursive = True);
        $this->json_property('payments','Payment', $many = True,
                             $recursive = True);
    }

    function process_for_transfer(){
        # can not update product by updating invoice
        # will have better solution later
        if($this->jobs){
            foreach($this->jobs as $job){
                $job->product = null;
            }
        }
    }

    function amount_paid(){
        #Return how much money has been paid for this invoice.
        if (is_array($this->payments)){
            $sum = 0;
            foreach($this->payments as $key=>$value){
                $sum += $value->amount;
            }
            return $sum;
        } else{
            return 0;
        }
    }

    function amount_owed(){
        #Return how much money still owed for this invoice
        $total = $this->total_cost;
        if($this->amount_paid()){
            return $total - $this->amount_paid();
        } else{
            return $total;
        }
    }

    function amount_owed_format_with_currency(){
        /*Return a formatted amount owed including the currency
            icon at the start of the value.
        */
        return format_currency($this->amount_owed(), 2, $this->currency);
    }

    function invoice_payment_status(){
        #Return the payment status of the invoice
        $total_payment = 0;
        foreach($this->payments as $key=>$value){
            $total_payment += $value->amount;
        }

        if ($total_payment < 0){
            return NEGATIVE_PAYMENT;
        } elseif ($total_payment == 0) {
            return NO_PAYMENT;
        } elseif ($this->total_cost == $total_payment) {
            return FULL_PAYMENT;
        } elseif ($this->total_cost > $total_payment) {
            return PART_PAYMENT;
        } else {
            return OVER_PAYMENT;
        }
    }

    function is_paid_string(){
        /*Return a string 'Unpaid' if the invoice is unpaid
          else return a string 'Paid'.
        */
        if ($this->unpaid){
            return "Unpaid";
        } else{
            return "Paid";
        }
    }
}

class Invoices extends Resource
{
    public function __construct()
    {
        $this->entity_class = 'Invoice';
        $this->json_name = 'invoices';
    }
}

$invoices = new Invoices();

File: python/invoices.py
import datetime
import sdk.python.entities
from sdk.python.payments import Payment
from sdk.python.items import Item
from sdk.python.addresses import Address
from sdk.python.stores import Store
from sdk.python.store_tags import StoreTag
from sdk.python.users import User
from sdk.python.files import File
from sdk.python.companies import Company
from sdk.python.phone_numbers import PhoneNumber
from sdk.python.email_addresses import EmailAddress
from sdk.python.util.money_protocol import format_currency
import sdk.python.util.invoice_status
from sdk.python.entities import Property


class Invoice(sdk.python.entities.Entity):

    resource = '/invoices/'
    json_name = 'invoice'

    note = Property(str)
    creation_date = Property(datetime.datetime)
    payment_deadline = Property(datetime.datetime)
    reminded = Property(datetime.datetime)
    archived = Property(datetime.datetime)
    id = Property(int)
    send_sms = Property(bool)
    send_email = Property(bool)
    unpaid = Property(bool)
    accept_stripe = Property(bool)
    accept_paypal = Property(bool)
    accept_bank_transfer = Property(bool)
    accept_phone_payment = Property(bool)
    is_remindable = Property(bool)
    force_reminders = Property(bool)
    total_cost = Property(float)
    subtotal_cost = Property(float)
    tax_amount = Property(float)
    currency = Property(str)
    invoice_token = Property(str)
    reminder_message = Property(str)
    responsible_manager = Property(User)
    creator = Property(User)
    client = Property(User)
    client_phone = Property(PhoneNumber)
    client_email = Property(EmailAddress)
    client_company = Property(Company)
    client_company_phone = Property(PhoneNumber)
    client_company_email = Property(EmailAddress)
    items = Property(Item, backref="invoice")
    shipping = Property(Address)
    store = Property(Store)
    pdf = Property(File)
    receipt = Property(File)
    payments = Property(Payment, backref="invoice")
    tags = Property(StoreTag, backref="invoices")

    def process_for_transfer(self):
        # can not update product by updating invoice
        # will have better solution later
        if self.jobs:
            for job in self.jobs:
                job.product = None

    def amount_paid(self):
        """ Return how much money has been paid for this invoice """
        if isinstance(self.payments, list):
            return sum([payment.amount for payment in self.payments])
        return 0

    def amount_owed(self):
        """ Return how much money still owed for this invoice """
        total = self.total_cost
        if self.amount_paid():
            return total - self.amount_paid()
        return total

    def amount_owed_formated_with_currency(self):
        """ Return a formatted amount owed including the currency
            icon at the start of the value.
        """
        return format_currency(self.amount_owed(), 2, self.currency)

    def invoice_payment_status(self):
        # This method returns the payment status of the invoice
        total_payments = \
            sum([payment.amount for payment in self.payments])
        if total_payments < 0:
            return sdk.python.util.invoice_status.NEGATIVE_PAYMENT
        elif total_payments == 0:
            return sdk.python.util.invoice_status.NO_PAYMENT
        elif self.total_cost == total_payments:
            return sdk.python.util.invoice_status.FULL_PAYMENT
        elif self.total_cost > total_payments:
            return sdk.python.util.invoice_status.PART_PAYMENT
        return sdk.python.util.invoice_status.OVER_PAYMENT

    def is_paid_string(self):
        """ Return a string 'Unpaid' if the invoice is unpaid
            else return a string 'Paid'.
        """
        if self.unpaid:
            return "Unpaid"
        return "Paid"


class Invoices(sdk.python.entities.Resource):

    entity_class = Invoice
    json_name = 'invoices'


invoices = Invoices()
